File: fix_audit.py
import re

filepath = '/home/team/shared/implementation_v1/src/data/breeds.ts'
with open(filepath, 'r') as f:
    content = f.read()

# 1. Fix "American Bull Dog" → "American Bulldog" - already done
# 2. Dobermann removal - already done
# 3. Doberman Pinscher merger - already done

# 4. Duplicate Hip Dysplasia in Akita (lines 175-180 approx)
# The second Hip Dysplasia entry in Akita:
#         {\n          "condition": "Hip Dysplasia",\n          "onsetAgeMonths": 18,\n          "severity": 2,\n          "screening": "OFA scoring"\n        },
# After Osteosarcoma entry, before Hypothyroidism entry
old_akita_hd = '''        {
          "condition": "Hip Dysplasia",
          "onsetAgeMonths": 18,
          "severity": 2,
          "screening": "OFA scoring"
        },
        {
          "condition": "Hypothyroidism",
          "onsetAgeMonths": 36,
          "severity": 2,
          "screening": "T4/TSH"
        }'''
new_akita = '''        {
          "condition": "Hypothyroidism",
          "onsetAgeMonths": 36,
          "severity": 2,
          "screening": "T4/TSH"
        }'''
content = content.replace(old_akita_hd, new_akita)
print("Fixed Akita duplicate Hip Dysplasia")

# 5. Alaskan Malamute - find the duplicate Hip Dysplasia
# First, let's find the Alaskan Malamute section
# Look for the second Hip Dysplasia entry by checking between sections
# Alaskan Malamute is at line 243 (after the rename and Dobermann removal)
# Let's find the pattern
am_pattern = r'"name": "Alaskan Malamute"[^}]*?predispositions": \[(.*?)\]'
am_match = re.search(am_pattern, content, re.DOTALL)
if am_match:
    preds = am_match.group(1)
    # Count Hip Dysplasia occurrences
    hd_count = len(re.findall(r'"condition": "Hip Dysplasia"', preds))
    print(f"Alaskan Malamute has {hd_count} Hip Dysplasia entries")
    if hd_count > 1:
        # Find the second occurrence - remove it
        # Split by patterns and remove the duplicate
        # Find the second Hip Dysplasia
        parts = preds.split('"condition": "Hip Dysplasia"')
        if len(parts) >= 3:
            # We have at least 2 occurrences
            # Find the block of the second one
            # The second occurrence starts after the first
            second_start = preds.index('"condition": "Hip Dysplasia"', 
                preds.index('"condition": "Hip Dysplasia"') + 1) - 1  # go back to find the {
            # Go back to find the opening {
            before_second = preds[:second_start]
            brace_start = before_second.rfind('{')
            if brace_start >= 0:
                # Find the closing }
                after_second = preds[second_start:]
                depth = 0
                end_pos = 0
                for i, c in enumerate(after_second):
                    if c == '{': depth += 1
                    if c == '}':
                        depth -= 1
                        if depth == 0:
                            end_pos = i + 1
                            break
                if end_pos > 0:
                    # Remove the block plus the preceeding comma
                    block = preds[brace_start:brace_start + end_pos]
                    # Check if there's a comma before
                    before = preds[:brace_start].rstrip()
                    if before.endswith(','):
                        block = ',' + block
                        brace_start -= 1
                    elif preds[brace_start + end_pos:brace_start + end_pos + 1] == ',':
                        end_pos += 1
                    
                    new_preds = preds[:brace_start] + preds[brace_start + end_pos:]
                    # Replace in content
                    old_full = f'"name": "Alaskan Malamute"[^}}]*?predispositions": [\\[].*?[\\]]'
                    content = content.replace(
                        f'"predispositions": [{preds}]',
                        f'"predispositions": [{new_preds}]'
                    )
                    print("Fixed Alaskan Malamute duplicate Hip Dysplasia")

# 6. Chow Chow - find and fix
cc_pattern = r'"name": "Chow Chow"[^}]*?predispositions": \[(.*?)\]'
cc_match = re.search(cc_pattern, content, re.DOTALL)
if cc_match:
    preds = cc_match.group(1)
    hd_count = len(re.findall(r'"condition": "Hip Dysplasia"', preds))
    print(f"Chow Chow has {hd_count} Hip Dysplasia entries")
    if hd_count > 1:
        # Remove second occurrence
        parts = preds.split('"condition": "Hip Dysplasia"')
        if len(parts) >= 3:
            second_start = preds.index('"condition": "Hip Dysplasia"', 
                preds.index('"condition": "Hip Dysplasia"') + 1) - 1
            before_second = preds[:second_start]
            brace_start = before_second.rfind('{')
            if brace_start >= 0:
                after_second = preds[second_start:]
                depth = 0
                end_pos = 0
                for i, c in enumerate(after_second):
                    if c == '{': depth += 1
                    if c == '}':
                        depth -= 1
                        if depth == 0:
                            end_pos = i + 1
                            break
                if end_pos > 0:
                    before = preds[:brace_start].rstrip()
                    if before.endswith(','):
                        brace_start -= 1
                    new_preds = preds[:brace_start] + preds[brace_start + end_pos:]
                    content = content.replace(
                        f'"predispositions": [{preds}]',
                        f'"predispositions": [{new_preds}]'
                    )
                    print("Fixed Chow Chow duplicate Hip Dysplasia")

# 7. Maine Coon - duplicate Obesity
mc_pattern = r'"name": "Maine Coon"[^}]*?predispositions": \[(.*?)\]'
mc_match = re.search(mc_pattern, content, re.DOTALL)
if mc_match:
    preds = mc_match.group(1)
    ob_count = len(re.findall(r'"condition": "Obesity"', preds))
    print(f"Maine Coon has {ob_count} Obesity entries")
    if ob_count > 1:
        parts = preds.split('"condition": "Obesity"')
        if len(parts) >= 3:
            second_start = preds.index('"condition": "Obesity"', 
                preds.index('"condition": "Obesity"') + 1) - 1
            before_second = preds[:second_start]
            brace_start = before_second.rfind('{')
            if brace_start >= 0:
                after_second = preds[second_start:]
                depth = 0
                end_pos = 0
                for i, c in enumerate(after_second):
                    if c == '{': depth += 1
                    if c == '}':
                        depth -= 1
                        if depth == 0:
                            end_pos = i + 1
                            break
                if end_pos > 0:
                    before = preds[:brace_start].rstrip()
                    if before.endswith(','):
                        brace_start -= 1
                    new_preds = preds[:brace_start] + preds[brace_start + end_pos:]
                    content = content.replace(
                        f'"predispositions": [{preds}]',
                        f'"predispositions": [{new_preds}]'
                    )
                    print("Fixed Maine Coon duplicate Obesity")

# 8. Add BOAS to French Bulldog, Boston Terrier, Pug, Shih Tzu
boas_block = '''        {
          "condition": "Brachycephalic Airway Syndrome",
          "onsetAgeMonths": 6,
          "severity": 4,
          "screening": "BOAS grading by veterinarian"
        }'''

boas_breeds = ['French Bulldog', 'Boston Terrier', 'Pug', 'Shih Tzu']

for breed_name in boas_breeds:
    # Find the breed's predispositions array
    pattern = r'"name": "' + re.escape(breed_name) + r'"[\s\S]*?"predispositions": \[(.*?)\]'
    match = re.search(pattern, content, re.DOTALL)
    if match:
        preds = match.group(1)
        if 'Brachycephalic Airway Syndrome' in preds:
            print(f"{breed_name} already has BOAS — skipping")
        else:
            # Add BOAS before the closing ]
            # Check if array is empty
            is_empty = len(preds.strip()) == 0
            prefix = '\n' + boas_block + '\n      ' if is_empty else ',\n' + boas_block + '\n      '
            new_preds = prefix + preds.lstrip()
            # Actually, simpler: insert before the closing ]
            old_full = f'"predispositions": [{preds}]'
            new_full = f'"predispositions": [{preds},\n{boas_block}\n      ]'
            content = content.replace(old_full, new_full, 1)
            print(f"Added BOAS to {breed_name}")

with open(filepath, 'w') as f:
    f.write(content)

print("\nDone! All fixes applied.")

File: fix_audit.mjs
import { readFileSync, writeFileSync } from 'fs';

const filePath = '/home/team/shared/implementation_v1/src/data/breeds.ts';
let content = readFileSync(filePath, 'utf-8');

// 1. Fix "American Bull Dog" → "American Bulldog"
content = content.replace('"name": "American Bull Dog"', '"name": "American Bulldog"');

// 2. Merge Dobermann into Doberman Pinscher
// The Dobermann block has predispositions: Hip Dysplasia, Elbow Dysplasia, GDV, Dilated Cardiomyopathy, Osteosarcoma, Hypothyroidism
// The Doberman Pinscher already has: Dilated Cardiomyopathy, Von Willebrand's, Hypothyroidism, Cervical Vertebral Instability
// We need to add the missing ones from Dobermann: Hip Dysplasia, Elbow Dysplasia, GDV, Osteosarcoma
// And upgrade DCM screening to the more specific one from Dobermann

// Add new predispositions to Doberman Pinscher's array (before the closing bracket of predispositions)
// The missing conditions from Dobermann: Hip Dysplasia, Elbow Dysplasia, GDV, Osteosarcoma
// DCM already exists (use the more detailed screening from Doberman Pinscher which is better)
// Hypothyroidism already exists

const dobermanPinscherPredispositions = [
  {
    condition: "Hip Dysplasia",
    onsetAgeMonths: 12,
    severity: 2,
    screening: "OFA scoring"
  },
  {
    condition: "Elbow Dysplasia",
    onsetAgeMonths: 12,
    severity: 2,
    screening: "OFA scoring"
  },
  {
    condition: "Gastric Dilatation-Volvulus",
    onsetAgeMonths: 12,
    severity: 4,
    screening: "Prophylactic gastropexy"
  },
  {
    condition: "Osteosarcoma",
    onsetAgeMonths: 48,
    severity: 3,
    screening: "Lameness evaluation"
  }
];

// Insert these after the last existing predisposition in Doberman Pinscher
// Find the Doberman Pinscher predispositions block
const dpPattern = `"name": "Doberman Pinscher",\\s*"species": "canine",\\s*"ideal_weight_min_kg": 30,\\s*"ideal_weight_max_kg": 45,\\s*"life_expectancy": 11,\\s*"senior_classification_age": 8,\\s*"alert_rules": \\{\\s*"predispositions": \\[`;
const dpMatch = content.match(new RegExp(dpPattern));
if (dpMatch) {
  // Find the end of the predispositions array in Doberman Pinscher
  // The array ends with the last entry before the closing ]
  const afterDp = content.slice(dpMatch.index + dpMatch[0].length);
  // Find the first ], that closes the predispositions array
  // We need to find the closing ] of the predispositions array
  let depth = 0;
  let endIdx = 0;
  for (let i = 0; i < afterDp.length; i++) {
    if (afterDp[i] === '[' || afterDp[i] === '{') depth++;
    if (afterDp[i] === ']' || afterDp[i] === '}') {
      depth--;
      if (depth === 0 && afterDp[i] === ']') {
        endIdx = i;
        break;
      }
    }
  }
  // Insert new predispositions just before the closing ]
  const insertPos = dpMatch.index + dpMatch[0].length + endIdx;
  const newEntries = dobermanPinscherPredispositions.map((p, i) => {
    const comma = i < dobermanPinscherPredispositions.length - 1 ? ',' : '';
    return `        {
          "condition": "${p.condition}",
          "onsetAgeMonths": ${p.onsetAgeMonths},
          "severity": ${p.severity},
          "screening": "${p.screening}"
        }${comma}`;
  }).join(',\n');
  content = content.slice(0, insertPos) + ',\n' + newEntries + content.slice(insertPos);
  console.log('Added Dobermann predispositions to Doberman Pinscher');
}

// Remove the Dobermann entry (lines 5210-5260)
// Match the full Dobermann block
const dobermannPattern = /  \{\s*\n\s*"name": "Dobermann",[\s\S]*?"contraindications": \[\]\s*\n\s*\}\s*\n\s*\},?\n\s*\{/;
const dobermannMatch = content.match(dobermannPattern);
if (dobermannMatch) {
  // Replace the Dobermann block + the following comma/brace
  const replacement = '  {';
  content = content.slice(0, dobermannMatch.index) + replacement + content.slice(dobermannMatch.index + dobermannMatch[0].length);
  console.log('Removed Dobermann duplicate entry');
}

// 3-4. Deduplicate Hip Dysplasia in Akita, Alaskan Malamute, Chow Chow
// And deduplicate Obesity in Maine Coon
const dedupTargets = [
  { breed: 'Akita', condition: 'Hip Dysplasia' },
  { breed: 'Alaskan Malamute', condition: 'Hip Dysplasia' },
  { breed: 'Chow Chow', condition: 'Hip Dysplasia' },
  { breed: 'Maine Coon', condition: 'Obesity' }
];

for (const target of dedupTargets) {
  // Find the breed entry
  const breedPattern = new RegExp(`"name": "${target.breed}"[\\s\\S]*?"predispositions": \\[`);
  const breedMatch = content.match(breedPattern);
  if (breedMatch) {
    const start = breedMatch.index + breedMatch[0].length;
    // Find the matching ] for predispositions
    let depth = 0;
    let endIdx = -1;
    const sub = content.slice(start);
    for (let i = 0; i < sub.length; i++) {
      if (sub[i] === '[' || sub[i] === '{') depth++;
      if (sub[i] === ']' || sub[i] === '}') {
        depth--;
        if (depth === 0 && sub[i] === ']') {
          endIdx = i;
          break;
        }
      }
    }
    if (endIdx > 0) {
      const predispositionsBlock = sub.slice(0, endIdx);
      // Find duplicates of the condition
      const condRegex = new RegExp(`"condition": "${target.condition}"`, 'g');
      const matches = [...predispositionsBlock.matchAll(condRegex)];
      if (matches.length > 1) {
        // Remove the duplicate entries (keep the first one)
        // Find the second occurrence and remove it
        const secondMatch = matches[1];
        // Find the full block of the second occurrence
        const blockStart = start + secondMatch.index;
        // Find the end of this block (the next }, or the end of the array)
        const afterSecond = predispositionsBlock.slice(secondMatch.index + secondMatch[0].length);
        // Go back to find the start of the block (find the preceding {)
        const beforeSecond = predispositionsBlock.slice(0, secondMatch.index);
        const lastOpenBrace = beforeSecond.lastIndexOf('{');
        const blockStartRelative = lastOpenBrace;
        
        // Find the matching } for this block
        let d = 0;
        let blockEnd = -1;
        for (let i = blockStartRelative; i < predispositionsBlock.length; i++) {
          if (predispositionsBlock[i] === '{') d++;
          if (predispositionsBlock[i] === '}') {
            d--;
            if (d === 0) {
              blockEnd = i;
              break;
            }
          }
        }
        if (blockEnd > 0) {
          // Remove from the start of the duplicate block to its end
          // Plus the comma after it
          let removeStart = start + blockStartRelative;
          let removeEnd = start + blockEnd + 1;
          // Remove the comma before the block (if not first entry)
          // Or the comma after it
          // Look for a leading comma
          const beforeChar = content.slice(removeStart - 1, removeStart);
          if (beforeChar === ',') {
            removeStart = removeStart - 1;
          } else {
            // Check for trailing comma
            const afterChar = content.slice(removeEnd, removeEnd + 1);
            if (afterChar === ',') {
              removeEnd = removeEnd + 1;
            }
          }
          content = content.slice(0, removeStart) + content.slice(removeEnd);
          console.log(`Removed duplicate "${target.condition}" in ${target.breed}`);
        }
      }
    }
  }
}

// 5. Add BOAS to French Bulldog, Boston Terrier, Pug, Shih Tzu
const boasEntry = `        {
          "condition": "Brachycephalic Airway Syndrome",
          "onsetAgeMonths": 6,
          "severity": 4,
          "screening": "BOAS grading by veterinarian"
        }`;

const boasBreeds = [
  { name: 'French Bulldog', add: true },
  { name: 'Boston Terrier', add: true },
  { name: 'Pug', add: null }, // check if already has it
  { name: 'Shih Tzu', add: null }  // check if already has it
];

for (const breed of boasBreeds) {
  const breedPattern = new RegExp(`"name": "${breed.name}"[\\s\\S]*?"predispositions": \\[`);
  const breedMatch = content.match(breedPattern);
  if (breedMatch) {
    const start = breedMatch.index + breedMatch[0].length;
    const sub = content.slice(start);
    let depth = 0;
    let endIdx = -1;
    for (let i = 0; i < sub.length; i++) {
      if (sub[i] === '[' || sub[i] === '{') depth++;
      if (sub[i] === ']' || sub[i] === '}') {
        depth--;
        if (depth === 0 && sub[i] === ']') {
          endIdx = i;
          break;
        }
      }
    }
    if (endIdx > 0) {
      const predispositionsBlock = sub.slice(0, endIdx);
      const hasBoas = predispositionsBlock.includes('Brachycephalic Airway Syndrome');
      
      if (breed.add === false) {
        // Skip if explicitly told not to add
      } else if (hasBoas) {
        console.log(`${breed.name} already has BOAS — skipping`);
      } else {
        // Insert BOAS before the closing ]
        const insertPos = start + endIdx;
        // Check if predispositions array is empty
        const isEmpty = predispositionsBlock.trim().length === 0;
        const prefix = isEmpty ? '\n' + boasEntry : ',\n' + boasEntry;
        content = content.slice(0, insertPos) + prefix + content.slice(insertPos);
        console.log(`Added BOAS to ${breed.name}`);
      }
    }
  }
}

writeFileSync(filePath, content, 'utf-8');
console.log('Done!');